File: backend/providers/openai_provider.py
from typing import AsyncIterator, Optional

from openai import AsyncOpenAI

from .base import LLMProvider


class OpenAIProvider(LLMProvider):
    name = "openai"

    def __init__(self, default_model: str = "gpt-4o"):
        self.default_model = default_model

    async def stream_vision_response(
        self,
        image_data_url: str,
        transcript_context: str,
        system_prompt: str,
        api_key: Optional[str] = None,
    ) -> AsyncIterator[str]:
        client = AsyncOpenAI(api_key=api_key)

        user_text = (
            f"Recent conversation transcript:\n{transcript_context}\n\n"
            "Here is the current screen. Respond with the most useful, "
            "concise assistance for what's happening right now."
        )

        stream = await client.chat.completions.create(
            model=self.default_model,
            stream=True,
            messages=[
                {"role": "system", "content": system_prompt},
                {
                    "role": "user",
                    "content": [
                        {"type": "text", "text": user_text},
                        {"type": "image_url", "image_url": {"url": image_data_url}},
                    ],
                },
            ],
        )
        async for chunk in stream:
            delta = chunk.choices[0].delta.content
            if delta:
                yield delta

File: android/app/src/main/java/com/wisp/app/providers/OpenAIProvider.kt
package com.wisp.app.providers

import kotlinx.coroutines.flow.Flow
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject

/**
 * Also the base for OpenRouterProvider, which is OpenAI-API-compatible and
 * differs only in base URL, model default, and header name for the key --
 * same relationship as the desktop app's openrouter_provider.py reusing
 * the `openai` Python SDK instead of writing a second HTTP client.
 */
open class OpenAIProvider(
    private val model: String = "gpt-4o",
    private val baseUrl: String = "https://api.openai.com/v1",
) : LLMProvider {
    override fun streamVisionResponse(
        imageBase64: String,
        mimeType: String,
        transcriptContext: String,
        systemPrompt: String,
        apiKey: String,
    ): Flow<String> {
        val userContent = JSONArray()
            .put(JSONObject().put("type", "text").put("text", buildUserText(transcriptContext)))
            .put(
                JSONObject()
                    .put("type", "image_url")
                    .put(
                        "image_url",
                        JSONObject().put("url", "data:$mimeType;base64,$imageBase64"),
                    ),
            )

        val body = JSONObject()
            .put("model", model)
            .put("stream", true)
            .put(
                "messages",
                JSONArray()
                    .put(JSONObject().put("role", "system").put("content", systemPrompt))
                    .put(JSONObject().put("role", "user").put("content", userContent)),
            )

        val request = Request.Builder()
            .url("$baseUrl/chat/completions")
            .addHeader("Authorization", "Bearer $apiKey")
            .post(body.toString().toRequestBody("application/json".toMediaType()))
            .build()

        return streamSse(request) { payload ->
            val choices = JSONObject(payload).optJSONArray("choices") ?: return@streamSse null
            if (choices.length() == 0) return@streamSse null
            val delta = choices.getJSONObject(0).optJSONObject("delta") ?: return@streamSse null
            delta.optString("content").takeIf { it.isNotEmpty() }
        }
    }
}
